src/samples/audio/smart_amp_test_ipc4.c

```diff
@@ -0,0 +1,449 @@
+	for (i = 0; i < frames; i++) {
+		for (j = 0 ; j < sad->out_channels; j++) {
+			if (chan_map[j] != -1) {
+				src = audio_stream_read_frag_s32(source,
+								 in_frag +
+								 chan_map[j]);
+				dest = audio_stream_write_frag_s32(sink,
+								   out_frag);
+				*dest = *src;
+			}
+			out_frag++;
+		}
+		in_frag += source->channels;
+	}
+	return 0;
+}
+
+static smart_amp_proc get_smart_amp_process(struct comp_dev *dev,
+					    struct comp_buffer __sparse_cache *buf)
+{
+	switch (buf->stream.frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		return smart_amp_process_s16;
+	case SOF_IPC_FRAME_S24_4LE:
+	case SOF_IPC_FRAME_S32_LE:
+		return smart_amp_process_s32;
+	default:
+		comp_err(dev, "smart_amp_process() error: not supported frame format");
+		return NULL;
+	}
+}
+
+/* Only reset the src_input and fb_input if the num_input_buffers
+ * (get from module adapter)has changed.
+ */
+static void smart_amp_input_setup(struct processing_module *mod,
+				  struct input_stream_buffer *inputs,
+				  int32_t num_input_buffers)
+{
+	struct smart_amp_data *sad = module_get_private_data(mod);
+	int i;
+
+	if (sad->input_stream_num != num_input_buffers) {
+		if (num_input_buffers == 1) {
+			sad->src_input = &inputs[0];
+			sad->fb_input = NULL;
+		} else {
+			for (i = 0; i < num_input_buffers; i++) {
+				if (IPC4_SINK_QUEUE_ID(inputs[i].id) ==
+				   SOF_SMART_AMP_FEEDBACK_QUEUE_ID)
+					sad->fb_input = &inputs[i];
+				else
+					sad->src_input = &inputs[i];
+			}
+		}
+
+		sad->input_stream_num = num_input_buffers;
+	}
+}
+
+static int smart_amp_process(struct processing_module *mod,
+			     struct input_stream_buffer *input_buffers, int num_input_buffers,
+			     struct output_stream_buffer *output_buffers, int num_output_buffers)
+{
+	struct smart_amp_data *sad = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct comp_buffer __sparse_cache *fb_buf;
+	struct comp_buffer __sparse_cache *src_buf;
+	uint32_t avail_passthrough_frames;
+	uint32_t avail_frames = 0;
+	uint32_t sink_bytes;
+	uint32_t fb_bytes;
+	uint32_t source_bytes;
+
+	smart_amp_input_setup(mod, input_buffers, num_input_buffers);
+	avail_passthrough_frames = sad->src_input->size;
+	if (sad->fb_input) {
+		fb_buf = attr_container_of(sad->fb_input->data,
+					   struct comp_buffer __sparse_cache,
+					   stream, __sparse_cache);
+
+		if (fb_buf->source && comp_get_state(dev, fb_buf->source) == dev->state) {
+			/* feedback */
+			avail_frames =
+				audio_stream_get_avail_frames(sad->fb_input->data);
+			avail_frames = MIN(avail_passthrough_frames,
+					   avail_frames);
+			fb_bytes = avail_frames *
+				audio_stream_frame_bytes(sad->fb_input->data);
+			buffer_stream_invalidate(fb_buf, fb_bytes);
+
+			sad->process(mod, sad->fb_input, &output_buffers[0],
+				     avail_frames, sad->config.feedback_ch_map);
+			sad->fb_input->consumed = fb_bytes;
+		}
+	}
+
+	if (!avail_frames)
+		avail_frames = avail_passthrough_frames;
```

Comment:
This could be put in an else clause, to save a boolean test.

Reply:
but add a assignment? Or it seems very strange.